funnel/models/membership.py

```diff
@@ -230,8 +248,68 @@ def replace(self, actor: User, **roles: Dict[str, Any]) -> ImmutableMembershipMi
         db.session.add(new)
         return new
 
+    def merge_and_replace(self, actor: User, other: MembershipType) -> MembershipType:
+        """Replace this record by merging roles from an independent record."""
+        if self.__class__ is not other.__class__:
+            # This should not be necessary if mypy catches incorrect calls, but it's
+            # also for safety from console scripting errors
+            raise TypeError("Merger requires membership records of the same type")
+        if other.revoked_at is not None:
+            raise MembershipRevokedError("Can't merge with a revoked membership record")
+
+        if (
+            self.record_type == MEMBERSHIP_RECORD_TYPE.INVITE
+            and other.record_type != MEMBERSHIP_RECORD_TYPE.INVITE
+        ):
+            # If we are an INVITE but the other is not an INVITE, then we must ACCEPT
+            # the INVITE before proceeding to an AMEND merger
+            this = self.accept(actor)
+        else:
+            # If both records are invites or neither is an invite, use existing records
+            this = self
+
+        role_columns = {}
+        for column in this.__data_columns__:
+            column_value = getattr(this, column)
+            if not column_value:
+                # Replace falsy values with value from the other record. This may need
+                # a more robust mechanism in the future if there are multi-value columns
+                column_value = getattr(other, column)
```

Comment:
What is `this` has a role that `other` does not? We keep the one from `this`?

Reply:
For boolean columns, yes. Behaviour is undefined for non-boolean columns. If we create one, we'll need to revise this method.